Fix filter by attribute widget now working for "Any..." attributes

Konamiman opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This pull requests fixes two bugs related to attribute filtering for variable products:

1. [Filter by attribute widget not counting variations having an attribute value of "Any..."](https://github.com/woocommerce/woocommerce/issues/27419)

When a variation has a value of "Any..." for a given attribute, then that variation isn't counted by the "Filter by attribute" widget corresponding to that attribute.

2. `attribute_` metadata entries with empty values not handled correctly when adding or removing variation attributes

This is an old bug (present in pre-4.4 versions). When a new variation attribute is added to the main product, then all the already existing variations will display the attribute with a value of "Any...", but the corresponding `attribute_` metadata entry for the variation won't be created. (If an existing variation is manually set to "Any..." and saved then it works as expected and the entry is created)

Conversely, when removing the attribute from the main product the corresponding `attribute_` metadata entries for the variations won't be removed.

A data migration is added too to backfill possible missing values for these metadata entries.

Closes #27419.

### How to test the changes in this Pull Request:

Start with an empty shop (no products).

1. Create an attribute named "Color" with three terms, e.g. "Black", "White", "Blue".
2. Create an attribute named "Style" with three terms, e.g. "Classic", "Sport", "Beach".
3. Add a "Filter by attribute" widget, to filter by "Color".
4. Add a variable product in which the "Color" attribute is used for all variations, with one variation per color.
5. Load the shop, verify that the filter by color widget is displaying all three colors, with a count of 1 for each.
6. Add a second variable product, but this time use the "Style" attribute to generate all variations.
7. After the three style variations have been created, add "Color" as an additional variation attribute with all three values. Now the existing variations should automatically get "Any..." as the value for the color attribute:

![image](https://user-images.githubusercontent.com/937723/91537309-5c945a00-e916-11ea-8fe3-792cd632eb49.png)

8. Reload the shop. Without the fix, it still displays 1 for all colors. With the fix, it displays 2 for each color (it's counting the "Any..." product now).
9. Set one of the variations to "Out of stock". Reload the shop and verify that the value of the counters hasn't changed.
10. Set all the variations to "Out of stock". Reload the shop and verify that the value of the counters is now 1 (and now the product doesn't display in the shop). Set the variations to "In stock" again.
11. Remove the "Color" attribute from the product, reload the shop and verify that all counters go back to 1.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Fix - Filter by attribute widget not working properly for variations having attribute values of "Any..."
> Fix - "attribute_" metadata entries not created and deleted properly for variations

Konamiman committed:
Fix filter by attribute widget now working for "Any..." attributes

When a variation product has an attribute with a value of "Any...",
and there's a filter by attribute widget for that attribute, then
that product won't be included in the counts displayed in the widget
(and if the count ends up being zero, the attribute won't be shown
in the widget).

This happens before since Woo 4.4, this widget works by looking at
entries in the term relationships table for varitions too
(used to do so for simple products and for "main" variable products
only), see PR #26260; but there are no such entries for
"Any..." attributes.

This commit fixes that by extending the SQL query used by the widget
to look for variations that have empty attribute values in the meta
table too.

Konamiman committed:
Update meta 'attribute_' values when variation attributes are added or removed.

When a new variation attribute is added, the corresponding 'attribute_'
meta entries are added for all variations with an empty value;
and when an existing variation attribute is removed, the existing
'attribute_' meta entries are deleted for all variations.

This is necessary for the filter by attribute widget to work properly
when variations exist with a value of "Any..." for attributes.

Konamiman committed:
Add data migration to backfill missing attributes meta for variations.

The previous commit fixes a bug that causes the "attribute_" metadata
with an empty value to not be created when a new variation attribute
is added to the product (so that all variations have the attribute
with a value of "Any..."). This commit adds a data migration to
backfill the missing metadata for existing variations.

rrennick commented:
@Konamiman You've got a unit test failure in PHP 7.4.

rrennick commented:
@Konamiman In the screenshot below the `Test 25133` product has 5 variations that are `Blue` so the count of `8` is technically correct. However, it's inconsistent with the number showing in the sorting and the number of products shown on the page.

<img width="1203" alt="Screen Shot 2020-08-28 at 10 22 37 AM" src="https://user-images.githubusercontent.com/343847/91565865-3b813880-e919-11ea-91f0-2aab4a6a24dd.png">

rrennick commented:
The other thing I wanted to test was adding an attribute term after the variable product had been created. I added `Extra Large`. 

<img width="905" alt="Screen Shot 2020-08-28 at 10 44 46 AM" src="https://user-images.githubusercontent.com/343847/91568629-f01c5980-e91b-11ea-96a6-253bdac9a749.png">

The first test product has `Any` size

<img width="1091" alt="Screen Shot 2020-08-28 at 10 45 13 AM" src="https://user-images.githubusercontent.com/343847/91568791-fd394880-e91b-11ea-8716-fa4448ee0553.png">

The widget on refresh

<img width="264" alt="Screen Shot 2020-08-28 at 10 46 13 AM" src="https://user-images.githubusercontent.com/343847/91568981-0a563780-e91c-11ea-9f06-f13fa9e51a75.png">

A hook needs to be added on updating attribute terms to sync the product lookup table.

Konamiman commented:
Hi @rrennick. After you add a new term you need to manually add that new term to the list of values used for variations here:

![image](https://user-images.githubusercontent.com/937723/91702488-1b4fb480-eb79-11ea-874c-ce68ecf42d63.png)

If what you mean is that adding a new term should automatically add the term to the list of attributes used for variations on all variable products using that attribute: maybe. But I think that's a separate discussion, not really related to this PR as that's the way it worked before.

Konamiman committed:
Change filter by attributes widget to count products, not variations.

Right now the filter by attributes widget counts available variations
(for variation products). This is confusing since the counter shows
numbers that are higher than the actual count of products displayed.

This commit changes the query used by the widget so that instead
of counting variations it returns the parent product ids, and then
counts the distinct values. This also covers the case of products
where some of the variations have concrete values and some have
"Any..." values.

Konamiman committed:
Update wc_product_attribute_uasort_comparison to accept null entries

When set_attributes is used on WC_Product to remove existing attributes
the wc_product_attribute_uasort_comparison ends up being called
with a null argument, and this breaks tests in PHP 7.4 since
null is used as an array. This commit modifies the function so that
if null is passed no array access is attempted.

vedanshujain commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -409,9 +410,36 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+		$query_sql = "
+			SELECT COUNT(DISTINCT(product_id)) AS term_count, term_count_id FROM (
+				{$main_query_sql}
+				UNION ALL
```
Instead of unionizing this, can we fetch the result in two separate queries and then combine. Basically group by over union is forcing MySQL to do a filesort with both the results, also this can have a large insert locking footprint.

I think we should favor performance here instead of atomicity.

ObliviousHarmony replied:
It's worth mentioning that we're caching this in a transient so if it's a little bit slower it isn't necessarily a significant problem so long as the locking footprint isn't against the posts table. @Konamiman and @vedanshujain what do you think of using this instead of a union or two queries:

```php
// Generate query.
$query           = array();
$query['select'] = "SELECT COUNT( DISTINCT {$wpdb->posts}.ID ) as term_count, terms.term_id as term_count_id";
$query['from']   = "FROM {$wpdb->posts}";
$query['join']   = "
	INNER JOIN {$wpdb->term_relationships} AS tr ON {$wpdb->posts}.ID = tr.object_id OR {$wpdb->posts}.post_parent = tr.object_id
	INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy ON tr.term_taxonomy_id = term_taxonomy.term_taxonomy_id
	LEFT JOIN (
		SELECT term_taxonomy_sub.taxonomy, tr_sub.object_id, term_taxonomy_sub.term_id FROM {$wpdb->term_relationships} AS tr_sub
		INNER JOIN {$wpdb->term_taxonomy} AS term_taxonomy_sub ON tr_sub.term_taxonomy_id = term_taxonomy_sub.term_taxonomy_id
	) AS tr_variation ON {$wpdb->posts}.ID = tr_variation.object_id AND term_taxonomy.taxonomy = tr_variation.taxonomy
	INNER JOIN {$wpdb->terms} AS terms
		ON tr_variation.term_id IS NULL AND term_taxonomy.term_id = terms.term_id
		OR tr_variation.term_id = terms.term_id
	{$main_tax_query_sql['join']} {$meta_query_sql['join']}"; // Not an omission, really no more JOINs required.
```

vedanshujain replied:
We discussed this offline, and the query suggested by @ObliviousHarmony is currently non-optimized, so going ahead with a different query added in https://github.com/woocommerce/woocommerce/pull/27508/commits/07b62dabbde388d17d9e7e772de4e8b29b10d410

vedanshujain reviewed (commented):
I am still reviewing and testing the PR, but adding some initial comments that will be helpful.

vedanshujain commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -365,9 +365,10 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+		// This one will return non-variable products and variable products with concrete values for the attributes.
 		$query           = array();
-		$query['select'] = "SELECT COUNT( DISTINCT {$wpdb->posts}.ID ) as term_count, terms.term_id as term_count_id";
+		$query['select'] = "SELECT {$wpdb->posts}.post_parent as product_id, terms.term_id as term_count_id";
```
are we sure about this change? `post_parent` is set to 0 for normal products, i think this is applicable only for variations.

vedanshujain commented on includes/wc-update-functions.php:
```diff
@@ -2242,3 +2242,41 @@ function wc_update_450_sanitize_coupons_code() {
+ *
+ * @return bool true if there are more products to process.
+ */
+function wc_update_450_insert_attribute_metadata_for_variations_not_having_it() {
```
any way to work around this migration? this is going to really really long time for many hosts, let's go an extra mile to avoid this if we can.

Konamiman replied:
But we are using those empty meta values to detect "Any" variations. If we don't ensure that these meta values exist for all variations, the widget won't work properly.

vedanshujain replied:
Resolving because we discussed offline and figured out a way to not need this migration.

ObliviousHarmony reviewed (changes requested):
Thanks for this @Konamiman, I've left some comments I think we should address.

Regarding the PR itself however I don't think this is the route that we should take. I'm worried that automatically synchronizing the attributes to variations is a little too destructive and might be considered backwards incompatible. Right now (without this PR), when you remove an attribute from a variable product, none of the terms are removed from the variations. This allows you to add back the attribute without needing to restore a backup.

It also appears that the behavior of the widget isn’t exactly the same as without the PR when things are handled correctly, but that may just be a misunderstanding on my part.

From what I can tell it seems like mostly everything other than this widget is resilient to the missing entries and so perhaps we should just correct the count query and leave the rest as it is for now?

Konamiman commented on includes/class-wc-product-variable.php:
```diff
@@ -468,10 +468,17 @@ public function save() {
 		$previous_name = $this->data['name'];
 		$new_name      = $this->get_name( 'edit' );
 
+		if ( $this->get_id() ) {
```
We need to sync empty meta values, because that's what we are using to count "Any" attributes.

vedanshujain replied:
Resolving this because we discussed offline and figured out a way to not need this sync.

Konamiman committed:
Improve the query to retrieve "Any..."-attributed variations in filtering widget

The new query doesn't need empty attribute entries in the meta table,
therefore the code that generates them and the migration to backfill
the missing existing ones have been removed.

ObliviousHarmony reviewed (approved):
I've tested this out with various product types and it seems to do what it says on the tin! 

One nitpick is that we're now counting variable products using the `post_parent` instead of their own `ID`. This behavior is different than what it was doing before as it was previously considering each in-stock variation in the count. This might lead to strange behavior when combining multiple widgets and I think we should use the `ID` for variations as well.

Konamiman commented:
> One nitpick is that we're now counting variable products using the post_parent instead of their own ID. This behavior is different than what it was doing before as it was previously considering each in-stock variation in the count. This might lead to strange behavior when combining multiple widgets and I think we should use the ID for variations as well.

Originally it did count variations, but as @rrennick pointed out, that resulted in the widget displaying a number that was higher than the actual amount of products being displayed, which was quite confusing. So it now looks at which variations are supposed to be visible, keeps its parent id, and counts how many different parent ids are in the result; this also blends nicely with the counts for simple products.

vedanshujain reviewed (changes requested):
I have left a comment about splitting the query, I found out a couple more issues but let's tackle them in different PR:
1. Looks like we have introduced performance regressions in `get_available_variations` and then calling it in `is_visible_core` for every product. This makes the whole page slow if there are many products with a large number of variations.
2. I am not sure if using `IF` in select statement and selecting `post_parent` is the best way because we are applying the rest of the checks to post.ID. to explain, if a variable product is marked out of stock, then the variations are still counted and shown if the particular variations are not marked out of stock. This is because the checking is done on the post.ID but we are bubbling up post.post_parent in the SELECT statement.

Let's deal with these two changes later on, but I think we should still split this query into two and run them separately.

Konamiman commented:
> if a variable product is marked out of stock, then the variations are still counted and shown if the particular variations are not marked out of stock

This is the point of the whole thing. Take in account only variations that are in stock, but in the end, get the main product id so that counters are consistent with the list of products shown (which are always the main products).

E.g. assuming one single product, if two variations are in stock for the same attribute, say "blue-big" and "blue-small", then count only one for "blue". Otherwise we would be displaying a count of 2 but there would be only one product in the displayed list, which would be confusing.

Or am I misunderstanding you?

vedanshujain commented:
> Take in account only variations that are in stock, but in the end, get the main product id so that counters are consistent with the list of products shown (which are always the main products).

So this becomes inconsistent with the main query when the parent product is out of stock but variations are in stock. Then we will count it in the widget, but main query won't as the product is tagged out of stock. That said, I mentioned that we will deal with this later because currently the only way to repro this scenario is to convert the product into simple, then make it out of stock, and then convert it back to the variable. This is definitely an edge case, let's not block the whole PR for this. (You can also manually add `out_of_stock` term relationship on a variable product to repro).

Konamiman committed:
Split the query for counting terms in the filter by attributes widget.

For performance reasons the query is split in two: one for simple
products and variations with a concrete attribute value, and another
one for variations having "Any..." as the attribute value.

vedanshujain commented on includes/widgets/class-wc-widget-layered-nav.php:
```diff
@@ -425,19 +442,53 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
+					{$query}
+				) AS x GROUP BY term_count_id";
+
+			$results = $wpdb->get_results( $query, ARRAY_A ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```
lets cache here instead of in callee function, so that both the queries can be cached separately

Konamiman replied:
Do we really need to cache the queries separately? We have split the original query in two for performance reasons, but conceptually they are still a single one; they will always be used together and with the same set of parameters.

vedanshujain reviewed (approved):
I have left a comment, but pre-approving since its minor. Note that this whole widget needs perf improvements, but I think we can merge it for now.

jonathansadowski commented:
Thanks for the reviews and work to get this in to 4.5

Gasper90 commented:
This is still not working on our side. The filter doesn't show up on the page.

Gasper90 commented:
Ok, because I saw this in the [changelog](https://wordpress.org/plugins/woocommerce/#developers): 

>  * Fix – Filter by attribute widget not working properly for variations having attribute values of “Any…”. #27508

jonathansadowski commented:
Sorry @Gasper90, I thought I was looking at the revert commit (#27625). You're correct. We originally included this PR in hopes that it would improve things, and have a revert queued for early next week.

Gasper90 commented:
It works now! 👍
